### ISSUE TITLE
Multiple boundary attributes cause parts to be missed completely

### ISSUE THREAD

**Issue by todd-richmond** [markov2/perl5-Mail-Message#20]
A serious issue has been discovered that if you supply multiple boundaries, Mail::Message will only parse content from the 1st boundary, but Apple Mail will parse the last boundary. That causes us to skip processing parts that the end-user sees in their mail client. I think the simplest fix is to always use the last boundary so that any content before that is considered prologue and ignored

Note that $field->attribute('boundary, '2') will only change the 1st boundary and not any subsequent boundaries if > 1 attribute

```
Date: Wed, 21 Aug 2024 10:10:49 +0000
MIME-Version: 1.0
Subject: a email contain virus
Content-Type: multipart/mixed; boundary=0;boundary=1;boundary=2

--1
Content-type: application/octet-stream; name=123
Content-Disposition: attachment; filename=noteicar

123
--1--
--2
Content-type: application/octet-stream; name=123
Content-Disposition: attachment; filename=eicar

X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*
--2--

**Comment by todd-richmond** [comment]
proposed solution
````
diff Multipart.pm.org Multipart.pm
448,449c448,449
<     {   my $boundary = $mime->attribute('boundary');
<         return $boundary if defined $boundary;
---
>     {   my @boundary = $mime->attributes('boundary');
>         return $boundary[-1] if $#boundary != -1;

**Comment by markov2** [comment]
I see no rule in RFC2049 which would make this a valid message... I think MailBox should refuse messages which are broken this way.  Why does Apple accept it?

**Comment by todd-richmond** [comment]
All mail clients are horribly "lenient" in their RFC interpretations which of course then leads to major security holes instead of requiring senders to send valid email. Outlook does all sorts of crazy stuff with "previews" and MIME parsing as well

**Comment by todd-richmond** [comment]
Confirmed that Outlook for Windows also shows the last mime part.

**Comment by markov2** [comment]
I am still very much divided how to tackle this issue.  It is a broken message, and do we break on those or pass them on?  Why is this really a problem for you?   My guess is that your spam-filter get a different idea about the content-type than Apple and Outlook.

Your suggested patch only accidentally works.  Method attributes() has no parameters and returns key-value pairs, so in the example case it returns `(boundary => 0, boundary => 1, boundary => 2)`.  The last value in this list is accidentally '2'.
You explain ` $field->attribute('boundary, '2')`, but this sets `boundary` to `2`.  Also not what you intended.

Probably, the Apple and Outlook code do something best mimic-ed by
```perl
my %attrs = $mime->attributes;    # the last use of the key remains
my $boundary = $attrs{boundary};
```
Isn't a rule which checks for an attempt to pass 'boundary' multiple times in one content-type not a nice indicator for a spam-filter?  Is this MailBox' task help in a work-around to support bad behavior?
```perl
my @a = $mime->attributes;
my @keys;
while(@a) { push @keys, shift @a; shift @a);  # is there an easier way to get even elements from a list?
$is_virus++ if grep($_ eq 'boundary', @keys) > 2;
```

**Comment by todd-richmond** [comment]
this is a major hole as given in the example. The 2nd attached content is not returned as a parsed mime part by mail::message so we never see it to scan it, but outlook and apple mail both allow users to download. We must scan each part for multiple different detection algorithms (not just spam and virus)

we must scan every part that an end user can see and or download. This simple patch allows mail::message to consider all content before the start of the final boundary to be preamble and ignored to better match what clients are doing - i.e. all it does is choose he last string instead of the 1st

**Comment by markov2** [comment]
You known that there also is a `$body->epilogue` for after the part termination?

For the solution, should we make it more general: that Outlook and Apple do always take the last of the attribute with a certain name?  F.i. 'filename'?

Should I (also) support multiple boundaries within a multipart: to convert it into a correct multipart structure?

Do you think this structure is specifically designed to exploit weaknesses in spamfilters?

**Comment by todd-richmond** [comment]
This is definitely designed to work around email filters so the content is attempting to do something horrible. We are doing our own mime structure rewriting for a variety of issues and so I've added code to rewrite the 1st bound to be same as the last (there can be many of boundaries, but only 1st and last are relevant

Always picking the last attribute seems fine since I haven't found a client that picks the 1st other than Mail::Message. I haven't tested filenames, but can try that tomorrow - probably last-value-wins though

Attempting to rewrite mime structure is going to be painful because what happens if > 2 boundaries. Since Outlook effectively sets it to be hidden prologue, we're fine skipping that content too since we only care about what clients can make clickable/downloadable

**Comment by todd-richmond** [comment]
looks great - thx! it's amazing how many "fix-ups" we have to do to work around mostly Outlook oddities

**Comment by markov2** [comment]
This change is released to CPAN, Mail::Message module version 3.016
